File: C++/source/Framework/Utils/FixedSizeBuffer.cpp
/***************************************************************************
 *   Copyright (C) 2005 Vidyasagara Guntaka                                *
 *   All rights reserved.                                                  *
 *   Author : Vidyasagara Reddy Guntaka                                    *
 ***************************************************************************/

#include "Framework/Utils/FixedSizeBuffer.h"
#include "Framework/Utils/TraceUtils.h"
#include "Framework/Utils/AssertUtils.h"

namespace WaveNs
{

FixedSizeBuffer::FixedSizeBuffer (const UI32 maximumSize)
    : m_maximumSize (maximumSize),
      m_isBufferOwnedByMe (true)
{
    if (0 == m_maximumSize)
    {
		trace (TRACE_LEVEL_FATAL, "FixedSizeBuffer::FixedSizeBuffer : A FixedSizeBuffer of length 0 is not possible.");
		waveAssert (false, __FILE__,__LINE__);
	}

    m_pRawBuffer  = new UI8[m_maximumSize + 1];
    m_currentSize = 0;

    if (NULL == m_pRawBuffer)
    {
        trace (TRACE_LEVEL_FATAL, "FixedSizeBuffer::FixedSizeBuffer : Could not allocate a buffer.");
        waveAssert (false, __FILE__,__LINE__);
    }

    m_pRawBuffer[m_maximumSize] = '\0';
}

FixedSizeBuffer::FixedSizeBuffer (const UI32 maximumSize, void *pBuffer, bool isBufferUsed)
    : m_maximumSize (maximumSize),
      m_isBufferOwnedByMe (false)
{
    if (0 == m_maximumSize)
    {
		trace (TRACE_LEVEL_FATAL, "FixedSizeBuffer::FixedSizeBuffer : A FixedSizeBuffer of length 0 is not possible.");
		waveAssert (false, __FILE__,__LINE__);
	}

    m_pRawBuffer  = reinterpret_cast<UI8 *> (pBuffer);

    if (true == isBufferUsed)
    {
        m_currentSize = m_maximumSize;
    }
    else
    {
        m_currentSize = 0;
    }

    if (NULL == m_pRawBuffer)
    {
        trace (TRACE_LEVEL_FATAL, "FixedSizeBuffer::FixedSizeBuffer : Invalid buffer supplied");
        waveAssert (false, __FILE__,__LINE__);
    }
}

FixedSizeBuffer::~FixedSizeBuffer ()
{
    if (true == m_isBufferOwnedByMe)
    {
        if (NULL != m_pRawBuffer)
        {
            delete[] m_pRawBuffer;
        }
    }

    m_pRawBuffer  = NULL;
    m_currentSize = 0;
}

UI8 *FixedSizeBuffer::getPRawBuffer ()
{
    return (m_pRawBuffer);
}

UI8 *FixedSizeBuffer::getPRawBuffer () const
{
    return (m_pRawBuffer);
}

UI8 *FixedSizeBuffer::getPCurrentRawBuffer ()
{
    return (m_pRawBuffer + m_currentSize);
}

const UI32 FixedSizeBuffer::getMaximumSize () const
{
    return (m_maximumSize);
}

const UI32 FixedSizeBuffer::getCurrentSize () const
{
    return (m_currentSize);
}

const UI32 FixedSizeBuffer::getRemainingSize () const
{
    if (m_currentSize > m_maximumSize)
    {
        trace (TRACE_LEVEL_FATAL, "FixedSizeBuffer::getRemainingSize : Current size exceeded Maximum Size.  Something went wrong.");
        waveAssert (false, __FILE__,__LINE__);
    }

    if (m_currentSize <= m_maximumSize)
    {
        return (m_maximumSize - m_currentSize);
    }
    else
    {
        return (0);
    }
}

void FixedSizeBuffer::incrementCurrentSize (const UI32 incrementSize)
{
    UI32 remainingSize = m_maximumSize - m_currentSize;

    if (remainingSize < incrementSize)
    {
        trace (TRACE_LEVEL_ERROR, "FixedSizeBuffer::incrementCurrentSize : Increment Size causes Current size to exceeded Maximum Size.  Something went wrong.");
        trace (TRACE_LEVEL_FATAL, "FixedSizeBuffer::incrementCurrentSize : Not incrementing the current size.");
        waveAssert (false, __FILE__,__LINE__);
    }
    else
    {
        m_currentSize += incrementSize;
    }
}

bool FixedSizeBuffer::isFull ()
{
    return (m_maximumSize == m_currentSize);
}

const void FixedSizeBuffer::toString (string &resultString)
{
    m_pRawBuffer[m_currentSize] = '\0';

    resultString.assign ((const char *) m_pRawBuffer, m_currentSize);
}

void  FixedSizeBuffer::reset ()
{
    m_currentSize = 0;
}

UI8 *FixedSizeBuffer::transferBufferToUser (UI32 &currentSize, UI32 &maximumSize)
{
    if (false == m_isBufferOwnedByMe)
    {
        waveAssert (false, __FILE__, __LINE__);
    }

    m_isBufferOwnedByMe = false;

    currentSize = m_currentSize;
    maximumSize = m_maximumSize;

    return (m_pRawBuffer);
}

}

File: Java/source/com/CxWave/Wave/Framework/Utils/Buffer/FixedSizeBuffer.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka & CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.Framework.Utils.Buffer;

import java.io.UnsupportedEncodingException;

import com.CxWave.Wave.Framework.Type.UI32;
import com.CxWave.Wave.Framework.Utils.Assert.WaveAssertUtils;
import com.CxWave.Wave.Framework.Utils.Trace.WaveTraceUtils;
import com.CxWave.Wave.Resources.ResourceEnums.TraceLevel;

public class FixedSizeBuffer
{
    private final UI32   m_maximumSize;
    private final byte[] m_rawBuffer;
    private UI32         m_currentSize;
    private boolean      m_isBufferOwnedByMe;

    public FixedSizeBuffer (final UI32 maximumSize)
    {
        m_maximumSize = maximumSize;
        m_isBufferOwnedByMe = true;

        if (0 == (m_maximumSize.getValue ()))
        {
            WaveTraceUtils.trace (TraceLevel.TRACE_LEVEL_FATAL, "FixedSizeBuffer.FixedSizeBuffer : A FixedSizeBuffer of length 0 is not possible.");
            WaveAssertUtils.waveAssert ();
        }

        m_rawBuffer = new byte[(m_maximumSize.intValue ()) + 1];
        m_currentSize = new UI32 (0);

        if (null == m_rawBuffer)
        {
            WaveTraceUtils.trace (TraceLevel.TRACE_LEVEL_FATAL, "FixedSizeBuffer.FixedSizeBuffer : Could not allocate a buffer.");
            WaveAssertUtils.waveAssert ();
        }

        m_rawBuffer[m_maximumSize.intValue ()] = '\0';
    }

    public FixedSizeBuffer (final UI32 maximumSize, final byte[] buffer, final boolean isBufferUsed)
    {
        m_maximumSize = maximumSize;
        m_isBufferOwnedByMe = false;

        if (0 == (m_maximumSize.getValue ()))
        {
            WaveTraceUtils.trace (TraceLevel.TRACE_LEVEL_FATAL, "FixedSizeBuffer.FixedSizeBuffer : A FixedSizeBuffer of length 0 is not possible.");
            WaveAssertUtils.waveAssert ();
        }

        m_rawBuffer = buffer;

        if (true == isBufferUsed)
        {
            m_currentSize = m_maximumSize;
        }
        else
        {
            m_currentSize = new UI32 (0);
        }

        if (null == m_rawBuffer)
        {
            WaveTraceUtils.trace (TraceLevel.TRACE_LEVEL_FATAL, "FixedSizeBuffer.FixedSizeBuffer : Could not allocate a buffer.");
            WaveAssertUtils.waveAssert ();
        }
    }

    public byte[] getRawBuffer ()
    {
        return (m_rawBuffer);
    }

    public UI32 getMaximumSize ()
    {
        return (m_maximumSize);
    }

    public UI32 getCurrentSize ()
    {
        return (m_currentSize);
    }

    public UI32 getRemainingSize ()
    {
        if ((m_currentSize.getValue ()) > (m_maximumSize.getValue ()))
        {
            WaveTraceUtils.trace (TraceLevel.TRACE_LEVEL_FATAL, "FixedSizeBuffer.getRemainingSize : Current size exceeded Maximum Size.  Something went wrong.");
            WaveAssertUtils.waveAssert ();
        }

        if ((m_currentSize.getValue ()) <= (m_maximumSize.getValue ()))
        {
            return (new UI32 ((m_maximumSize.getValue ()) - (m_currentSize.getValue ())));
        }
        else
        {
            return (new UI32 (0));
        }
    }

    public void incrementCurrentSize (final UI32 incrementSize)
    {
        final UI32 remainingSize = getRemainingSize ();

        if ((incrementSize.getValue ()) > (remainingSize.getValue ()))
        {
            WaveTraceUtils.trace (TraceLevel.TRACE_LEVEL_ERROR, "FixedSizeBuffer.incrementCurrentSize : Increment Size causes Current size to exceeded Maximum Size.  Something went wrong.");
            WaveTraceUtils.trace (TraceLevel.TRACE_LEVEL_FATAL, "FixedSizeBuffer.incrementCurrentSize : Not incrementing the current size.");
            WaveAssertUtils.waveAssert ();
        }
        else
        {
            m_currentSize.increment (incrementSize);
        }
    }

    public boolean isFull ()
    {
        return (0 >= (getRemainingSize ()).getValue ());
    }

    @Override
    public String toString ()
    {
        String newString = null;

        try
        {
            newString = new String (m_rawBuffer, "UTF-8");
        }
        catch (final UnsupportedEncodingException e)
        {
            WaveTraceUtils.tracePrintf (TraceLevel.TRACE_LEVEL_FATAL, "FixedSizeBuffer.toString : Unsupported Encoding : %s", "UTF-8");
            WaveAssertUtils.waveAssert ();
        }

        return (newString);
    }

    public void reset ()
    {
        m_currentSize = new UI32 (0);
    }

    public byte[] transferBufferToUser (final UI32 currentSize, final UI32 maximumSize)
    {
        if (false == m_isBufferOwnedByMe)
        {
            WaveAssertUtils.waveAssert ();
        }

        m_isBufferOwnedByMe = false;

        currentSize.setValue (m_currentSize.getValue ());
        maximumSize.setValue (m_maximumSize.getValue ());

        return (m_rawBuffer);
    }
}
